wgengine/userspace.go

```diff
@@ -360,22 +368,50 @@ func (e *userspaceEngine) isLocalAddr(ip packet.IP) bool {
 
 // handleDNS is an outbound pre-filter resolving Tailscale domains.
 func (e *userspaceEngine) handleDNS(p *packet.ParsedPacket, t *tstun.TUN) filter.Response {
-	if e.resolver.AcceptsPacket(p) {
-		// TODO(dmytro): avoid this allocation without having tsdns know tstun quirks.
-		buf := make([]byte, tstun.MaxPacketSize)
-		offset := tstun.PacketStartOffset
-		response, err := e.resolver.Respond(p, buf[offset:])
-		if err != nil {
-			e.logf("DNS resolver error: %v", err)
-		} else {
-			t.InjectInboundDirect(buf[:offset+len(response)], offset)
+	if p.DstIP == magicDNSIP && p.DstPort == magicDNSPort && p.IPProto == packet.UDP {
+		request := tsdns.Packet{
+			SrcIP:   uint32(p.SrcIP),
+			SrcPort: p.SrcPort,
+			Payload: p.Payload(),
 		}
-		// We already handled it, stop.
+		e.resolver.EnqueueRequest(request)
 		return filter.Drop
 	}
 	return filter.Accept
 }
 
+// pollResolver reads responses from the DNS resolver and injects them inbound.
+func (e *userspaceEngine) pollResolver() {
+	for {
+		resp, err := e.resolver.NextResponse()
+		if err == tsdns.ErrClosed {
+			return
+		}
+		if err != nil {
+			e.logf("tsdns: error: %v", err)
+			continue
+		}
+
+		const offset = tstun.PacketStartOffset
+		h := packet.UDPHeader{
+			IPHeader: packet.IPHeader{
+				SrcIP: packet.IP(magicDNSIP),
+				DstIP: packet.IP(resp.SrcIP),
```

Comment:
what if added a method on netaddr.IP to return the packed uint32 and documented it's only valid for IPv4 addresses?

Reply:
That would work great it you're OK with it.

Reply:
I was hesitant to propose stuff there that isn't likely to be widely useful to avoid bloating it up too much.